### PR TITLE
Make sure the noisy matrices are noninvertible

### PR DESCRIPTION
**Pull request by mbenke**

### CODE REVIEW

**Comment by kubkon** [comment]
For the future, it'd be good if you could specify whether any PR builds on any other. For instance, it seems that this one builds on #8, so naturally, we should first work on #8 before touching this PR.

File: src/algebra/m231.rs [hunk 1]
```diff
@@ -229,6 +230,17 @@ impl PartialEq<u32> for Mod231 {
     }
 }
 
+/// Try to solve the equation x^2 = a in our modular arithmetic
+/// Since MODULUS = 3 (mod 4), the solution, if exists, is x = a^((MODULUS+1)/4)
+pub fn try_sqrt(a: Mod231) -> Option<Mod231> {
+    let x = num_traits::pow::pow(a, MAGIC229);
```
Comment (kubkon) [inline src/algebra/m231.rs]: I was wondering, could we use [checked_pow] instead perhaps?

[checked_pow]: https://doc.rust-lang.org/std/primitive.u32.html#method.checked_pow

Comment (mbenke) [reply src/algebra/m231.rs]: no, we don't want to check for overflow, as we are using modular arithmetic

File: src/algebra/m231.rs [hunk 2]
```diff
@@ -7,10 +7,11 @@ use serde::{Deserialize, Serialize};
 use std::fmt;
 use std::ops::{Add, AddAssign, Div, Mul, MulAssign, Neg, Sub, SubAssign};
 
-const MODULUS: u32 = 2147483647u32;
+const MODULUS: u32 = 2147483647u32; // 2^31 -1
 const MODULUSI32: i32 = 2147483647i32;
 const MODULUSI64: i64 = 2147483647i64;
 const MODULUSU64: u64 = 2147483647u64;
+const MAGIC229: usize = 536870912; // (MODULUS+1) / 4 = 2^29
```
[the remark this answers is not in the document]
Comment (mbenke) [reply src/algebra/m231.rs]: added comment referring to `try_sqrt`

**Comment by mbenke** [comment]
> For the future, it'd be good if you could specify whether any PR builds on any other. For instance, it seems that this one builds on #8, so naturally, we should first work on #8 before touching this PR.

This PR is independent - can be applied on master as well as #8

**Comment by kubkon** [comment]
Since #10 landed, I'll close this PR as it covers the same ground but with #8 as a dependency. If you feel I missed something though, please feel free to reopen this PR.

**Pull request closed without merging**